### PR TITLE
Added the ability to upload a zip via the CLI

### PR DESCRIPTION
**Pull request by Graeme-Miller**
@geomacy would you please review?

### CODE REVIEW

**Comment by Graeme-Miller** [comment]
@geomacy can you review now you are back? cheers

File: cli/api/catalog/catalog.go [hunk 1]
```diff
@@ -23,6 +23,14 @@ import (
 	"fmt"
 	"github.com/apache/brooklyn-client/cli/models"
 	"github.com/apache/brooklyn-client/cli/net"
+	"net/url"
+        "path/filepath"
```
Comment (johnmccabe) [inline cli/api/catalog/catalog.go]: indentation, worth adding `golint` and `go vet` to the build process to pick up stuff like this.

**Comment by geomacy** [comment]
Code changes look good; but problems with test:

I tested this out as follows:
- upload a zip file as `br catalog add test.zip` - success
- upload a directory as `br catalog add test4` - success
- upload a zip file via "file" URL: `br catalog add file:///Users/geoff/tmp/test3/test.zip` - fails, with
open file:/Users/geoff/tmp/test3/test.zip: no such file or directory

From looking at the code I thought this should work, is this how it's intended?

**Comment by ahgittin** [comment]
confirmed that file URLs do not seem to work.  additionally i tried the following `https` url:

`target/br catalog add 'https://doc-08-bs-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/gvfvaob3c31nd2c4sq8j2d7rra0uogbf/1493028000000/05557496147892979345/*/0B3XurVLRa7pIWnhRV0swVktreUU?e=download'`

and got:

`ReaderException: special characters are not allowed`

**Comment by ahgittin** [comment]
also noted that:

```
% target/br catalog add 
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x20 pc=0x12c9794]

goroutine 1 [running]:
github.com/apache/brooklyn-client/cli/api/catalog.ZipResource.func1(0x0, 0x0, 0x0, 0x0, 0x150a540, 0xc420011860, 0x20, 0x1355c80)
	/Users/alex/go/src/github.com/apache/brooklyn-client/cli/api/catalog/catalog.go:188 +0x44
```

**Comment by ahgittin** [comment]
I also noted that any update that doesn't change a version is silently ignored, no error and no effect.  desired behaviour is that:

* if the update is a snapshot, simply replace it in the catalog
* if the update is not a snapshot, generate an error

This is NOT to do with this PR; it needs a change in the brooklyn-server REST API behaviour -- but I only noticed it here.  (I was running with https://github.com/apache/brooklyn-server/pull/645 when observed but I don't think that is significant.)

The idea is that a workflow where you make a change to a BOM or script, then `br catalog add` and good things happen if you're on a snapshot version -- do another `br deploy` and you should get the latest version.  (And ideally we could then add an `--update-all-deployments` to `catalog add` and deployments would be updated!)

Already with this PR it is much easier you just have to remember to bump the version number but we can do better.

**Comment by ahgittin** [comment]
@Graeme-Miller @geomacy To summarise above comments, I think it is important that we:

* ensure `file:` URLs work, at least for files (directories optional)
* ensure `http` and `https` URLs work, at least for files (directories impossible to guarantee; heuristics might be nice but not worth it), including the link above which gave the `ReaderException`
* update snapshot items in catalog (separate PR obviously)

and would be nice to:

* display nice error message if incomplete input supplied

We could merge this PR as is if there is not immediate bandwidth as the feature is already _so_ useful -- I just wouldn't want the observed issues to get forgotten.

**Comment by ahgittin** [comment]
working on fixing the items noted above -- in https://github.com/ahgittin/brooklyn-client/commit/6f6e73655af4b51abf2eaa747d3f343673e5e7a8

will open a new PR including that

**Comment by ahgittin** [comment]
I've merged this, as it's useful, and opened PR #48 fixing the other items noted.  It addresses everything Geoff and I listed, except for the snapshot-not-updating problem which is server side (and I'll look at next).

**Pull request merged**